halo/evmengine/keeper/keeper_internal_test.go

```diff
@@ -0,0 +1,177 @@
+			wantErr:    false,
+		},
+		{
+			name: "proposer false",
+			args: args{
+				header:  cmtproto.Header{Height: 1},
+				current: 0,
+				next:    2,
+			},
+			want:       false,
+			wantHeight: 2,
+			wantErr:    false,
+		},
+		{
+			name: "validatorsFunc error",
+			args: args{
+				header:  cmtproto.Header{Height: 1},
+				current: 0,
+				next:    1,
+				validatorsFunc: func(ctx context.Context, i int64) (*cmttypes.ValidatorSet, bool, error) {
+					return nil, false, errors.New("error")
+				},
+			},
+			want:    false,
+			wantErr: true,
+		},
+		{
+			name: "validatorsFunc not ok",
+			args: args{
+				header:  cmtproto.Header{Height: 1},
+				current: 0,
+				next:    1,
+				validatorsFunc: func(ctx context.Context, i int64) (*cmttypes.ValidatorSet, bool, error) {
+					return nil, false, nil
+				},
+			},
+			want:    false,
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			cdc := getCodec(t)
+			txConfig := authtx.NewTxConfig(cdc, nil)
+			mockEngine, err := newMockEngineAPI()
+			require.NoError(t, err)
+			cmtAPI := newMockCometAPI(t, tt.args.validatorsFunc)
+			tt.args.header.ProposerAddress = cmtAPI.validatorSet.Validators[tt.args.current].Address
+			nxtAddr, err := k1util.PubKeyToAddress(cmtAPI.validatorSet.Validators[tt.args.next].PubKey)
+			require.NoError(t, err)
+			ctx, storeService := setupCtxStore(t, tt.args.header)
+			keeper := NewKeeper(cdc, storeService, &mockEngine, txConfig)
+			keeper.cmtAPI = cmtAPI
+			keeper.addrProvider = mockAddressProvider{
+				address: nxtAddr,
+			}
+
+			got, gotHeight, err := keeper.isNextProposer(ctx)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("isNextProposer() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
+				t.Errorf("isNextProposer() got = %v, want %v", got, tt.want)
+			}
+			if gotHeight != tt.wantHeight {
+				t.Errorf("isNextProposer() gotHeight = %v, want %v", gotHeight, tt.wantHeight)
+			}
+		})
+	}
+}
+
+type mockCometAPI struct {
+	comet.API
+	fuzzer         *fuzz.Fuzzer
+	validatorSet   *cmttypes.ValidatorSet
+	validatorsFunc func(context.Context, int64) (*cmttypes.ValidatorSet, bool, error)
+}
+
+func newMockCometAPI(t *testing.T, valFun func(context.Context, int64) (*cmttypes.ValidatorSet, bool, error)) mockCometAPI {
+	t.Helper()
+	fuzzer := newFuzzer(0)
+	valSet := fuzzValidators(t, fuzzer)
+
+	return mockCometAPI{
+		fuzzer:         fuzzer,
+		validatorSet:   valSet,
+		validatorsFunc: valFun,
+	}
+}
+
+func fuzzValidators(t *testing.T, fuzzer *fuzz.Fuzzer) *cmttypes.ValidatorSet {
+	t.Helper()
+	var validators []*cmttypes.Validator
+
+	fuzzer.NilChance(0).NumElements(3, 7).Fuzz(&validators)
+
+	valSet := new(cmttypes.ValidatorSet)
+	err := valSet.UpdateWithChangeSet(validators)
+	require.NoError(t, err)
+
+	return valSet
+}
+
+func (m mockCometAPI) Validators(ctx context.Context, height int64) (*cmttypes.ValidatorSet, bool, error) {
```

Comment:
you need to assert `isNextProposer` uses the correct `height` when querying here. If we change the logic to `header.Height + 999` this test will still pass.